pkg/apis/v1alpha2/serviceimport.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+
+// ServiceImport describes a service imported from clusters in a ClusterSet.
+type ServiceImport struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// spec defines the behavior of a ServiceImport.
+	// +optional
+	Spec ServiceImportSpec `json:"spec,omitempty"`
+	// status contains information about the exported services that form
+	// the multi-cluster service referenced by this ServiceImport.
+	// +optional
+	Status ServiceImportStatus `json:"status,omitempty"`
+}
+
+// ServiceImportType designates the type of a ServiceImport
+type ServiceImportType string
+
+const (
+	// ClusterSetIP services are only accessible via the ClusterSet IP.
+	ClusterSetIP ServiceImportType = "ClusterSetIP"
+	// Headless services allow backend pods to be addressed directly.
+	Headless ServiceImportType = "Headless"
+)
+
+// ServiceImportSpec describes an imported service and the information necessary to consume it.
+type ServiceImportSpec struct{}
+
+// ServiceImportStatus describes derived state of an imported service.
+type ServiceImportStatus struct {
+	// +listType=atomic
+	Ports []v1.ServicePort `json:"ports"`
+	// ips will be used as the VIP(s) for this service when type is ClusterSetIP.
+	// +kubebuilder:validation:MaxItems:=1
+	// +optional
+	IPs []string `json:"ips,omitempty"`
+	// type defines the type of this service.
+	// Must be "ClusterSetIP" or "Headless".
+	// The "ClusterSetIP" type reflects exported Service(s) with type ClusterIP
+	// and the "Headless" type reflects exported Service(s) with type Headless.
+	// A ServiceImport with type ClusterSetIP SHOULD populate `.status.ips`
+	// with virtual IP address(es) where the service can be reached from within
+	// the importing cluster.
+	// If exported Services of the same name and namespace in a given ClusterSet
+	// have differing types, a "Conflict" status condition SHOULD be reported in
+	// ServiceExport status.
+	// +kubebuilder:validation:Enum=ClusterSetIP;Headless
+	Type ServiceImportType `json:"type"`
```

Comment:
The docs here are my expectation of how this is expected to work. Would appreciate feedback if this is inaccurate, e.g. if it should actually be possible to create a ClusterSetIP type ServiceImport for underlying exported Headless type Services.